nflow-engine/src/main/java/io/nflow/engine/internal/dao/WorkflowInstanceDao.java

```diff
@@ -867,7 +872,11 @@ public boolean setSignal(long workflowInstanceId, Optional<Integer> signal, Stri
   }
 
   public String getWorkflowInstanceType(long workflowInstanceId) {
-    return jdbc.queryForObject("select type from nflow_workflow where id = ?", String.class, workflowInstanceId);
+    String type = workflowTypeByWorkflowIdCache.computeIfAbsent(workflowInstanceId, id -> jdbc.queryForObject("select type from nflow_workflow where id = ?", String.class, id).intern());
```

Comment:
Can intern() cause memory leaks here?

Reply:
That was a problem in Java 7 and older, but we only support 8+, so not really. Also there will not be that many workflow types and they are not very large strings anyway.